spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPartitioningAwareScan.java

```diff
@@ -0,0 +1,267 @@
+  private StructLikeSet groupingKeys = null; // lazy cache of grouping keys
+
+  SparkPartitioningAwareScan(
+      SparkSession spark,
+      Table table,
+      Scan<?, ? extends ScanTask, ? extends ScanTaskGroup<?>> scan,
+      SparkReadConf readConf,
+      Schema expectedSchema,
+      List<Expression> filters) {
+
+    super(spark, table, readConf, expectedSchema, filters);
+
+    this.scan = scan;
+    this.preserveDataGrouping = readConf.preserveDataGrouping();
+
+    if (scan == null) {
+      this.specs = Collections.emptySet();
+      this.tasks = Collections.emptyList();
+      this.taskGroups = Collections.emptyList();
+    }
+  }
+
+  protected abstract Class<T> taskJavaClass();
+
+  protected Scan<?, ? extends ScanTask, ? extends ScanTaskGroup<?>> scan() {
+    return scan;
+  }
+
+  @Override
+  public Partitioning outputPartitioning() {
+    if (groupingKeyType().fields().isEmpty()) {
+      LOG.info("Reporting UnknownPartitioning with {} partition(s)", taskGroups().size());
+      return new UnknownPartitioning(taskGroups().size());
+    } else {
+      LOG.info(
+          "Reporting KeyGroupedPartitioning by {} with {} partition(s)",
+          groupingKeyTransforms(),
+          taskGroups().size());
+      return new KeyGroupedPartitioning(groupingKeyTransforms(), taskGroups().size());
+    }
+  }
+
+  @Override
+  protected StructType groupingKeyType() {
+    if (groupingKeyType == null) {
+      if (preserveDataGrouping) {
+        this.groupingKeyType = computeGroupingKeyType();
+      } else {
+        this.groupingKeyType = StructType.of();
+      }
+    }
+
+    return groupingKeyType;
+  }
+
+  private StructType computeGroupingKeyType() {
+    return org.apache.iceberg.Partitioning.groupingKeyType(expectedSchema(), specs());
+  }
+
+  private Transform[] groupingKeyTransforms() {
+    if (groupingKeyTransforms == null) {
+      List<PartitionField> groupingKeyFields = Lists.newArrayList();
+
+      Set<Integer> seenFieldIds = Sets.newHashSet();
+
+      for (PartitionSpec spec : specs()) {
+        for (PartitionField field : spec.fields()) {
+          int fieldId = field.fieldId();
+
+          if (groupingKeyType().containsField(fieldId) && !seenFieldIds.contains(fieldId)) {
+            groupingKeyFields.add(field);
+          }
+
+          seenFieldIds.add(fieldId);
+        }
+      }
+
+      this.groupingKeyTransforms = Spark3Util.toTransforms(table().schema(), groupingKeyFields);
+    }
+
+    return groupingKeyTransforms;
+  }
+
+  protected Set<PartitionSpec> specs() {
+    if (specs == null) {
+      Set<PartitionSpec> taskSpecs = Sets.newHashSet();
+      for (T task : tasks()) {
+        taskSpecs.add(task.spec());
+      }
+      this.specs = taskSpecs;
+    }
+
+    return specs;
+  }
+
+  protected synchronized List<T> tasks() {
+    if (tasks == null) {
+      try (CloseableIterable<? extends ScanTask> taskIterable = scan.planFiles()) {
+        List<T> plannedTasks = Lists.newArrayList();
```

Comment:
Minor: It looks like this was already done in `SparkBatchQueryScan`, but I generally prefer to keep things as a `CloseableIterable` as long as possible. This could probably return a `CloseableIterable` instead.

Reply:
I agree with you but I am not sure it applies here. We eagerly loaded tasks from day 1. They are used in lots of places and will need to be materialized immediately to compute stats, determine grouping key, etc.

I think we will be able to gc filtered out tasks in the current implementation as well.